File: modem_acustico/transmissor.m
function [y, msg_encoded] = transmissor(msg, RB, Fp, Fa)

%%%%%%%%%%%%%%%%%%%%%%  Entrada do transmissor   %%%%%%%%%%%%%%%%
%   msg           -- String (texto) a ser transmitida
%   RB            -- Taxa de bits (Bits/s)
%   Fp            -- Frequencia da portadora (Hz)
%   Fa            -- Frequencia de amostragem da placa de som (Hz)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Saida:
%   y             -- Sinal modulado pronto para transmissão
%   msg_encoded   -- Vetor de bits final após enquadramento e codificação
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% Detectar ambiente (Octave ou MATLAB)
% Assume-se Octave se não for MATLAB (para fins de carregamento de pacotes)
IS_OCTAVE = exist('OCTAVE_VERSION', 'builtin') ~= 0;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% Pacotes de software
if IS_OCTAVE
    pkg load communications;
    pkg load signal; % Para rcosine
end

%% Definições de Enquadramento
SFD = [1 1 0 0 1 1 1 0 0 0 1 1 1 1 0 0 0 0 1 1 1 0 0 0 1 1 0 0 1 0 1 0]; % 32 bits
% PREAMBLE: 10 unsynchronized bits, followed by 15 symbol intervals of 1010...
PRE_UNSYNC = ones(1, 10);
PRE_SYNC_PATTERN = upsample([1 0 1 0 1 0 1 0 1 0 1 0 1 0 1], 2); % 30 bits (15 símbolos x 2)

% Se o preâmbulo é usado apenas para a detecção de energia e sincronização de clock,
% o original pode ser simplificado (original: [ones(1,10) upsample(ones(1,15),2)]).
% Adotando a sequência original, mas renomeando para clareza:
PREAMBLE = [ones(1, 10) upsample(ones(1, 15), 2)];

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% 1. Conversão de Texto para Bits e Codificação
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
if ischar(msg) && ~isempty(msg)
    l = length(msg);                  % Comprimento da mensagem em bytes
    msg_dec = double(msg);            % Converte string para ASCII (decimal)

    % Monta o quadro de dados (Byte de Comprimento + Dados ASCII)
    % A função de2bi/dec2bin é usada para converter para bits (8 por caractere).
    % Assume-se a ordem de bits necessária para o receptor (geralmente LSB ou MSB no último).
    % Usando 'right-msb' (MSB da direita/primeiro) para compatibilidade:
    
    % Byte de Comprimento (l)
    l_bits = de2bi(l, 8, 'right-msb'); % Byte de comprimento como vetor [1x8]

    % Dados da Mensagem (msg_dec)
    msg_data_bits = reshape(de2bi(msg_dec, 8, 'right-msb')', 1, []);

    % Vetor de bits antes da codificação: [Byte_Comprimento | Dados]
    msg_bits_raw = [l_bits msg_data_bits];

    % Codificação de canal (Hamming)
    msg_encoded = hamming_encode(msg_bits_raw);
else
    % Caso de teste ou input não-string (mantém a compatibilidade com o caso MATLAB original)
    if isnumeric(msg)
       msg_encoded = msg;
    else
       error('Input "msg" deve ser uma string de texto.');
    end
end

disp(['Tamanho do payload (bits) antes do enquadramento: ' num2str(length(msg_encoded))]);


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% 2. Enquadramento                           %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% [PREAMBLE | SFD | PAYLOAD | PREAMBLE]
msg_framed = [PREAMBLE SFD msg_encoded PREAMBLE];


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% 3. Codificacao Polar (Banda Base)            %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Codificação polar: 0 -> -1, 1 -> +1
s = [-1, 1];
y_polar = s(msg_framed + 1);

disp(['Tamanho do quadro: ' num2str(length(y_polar)) ' símbolos']);


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% 4. Filtragem para Formatação de Pulso (RC/RRC) %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
r = 0.5;                             % Fator de decaimento (roll-off)
sps = floor(Fa / RB);                % Amostras por Símbolo

if IS_OCTAVE
    % Ajuste da taxa de bits para garantir um número inteiro de amostras/símbolo
    RB_f = Fa / sps;
    
    % Projeto do filtro Raised-Cosine (Octave: usa rcosine)
    num = rcosine(RB_f, Fa, 'default', r); % Coeficientes do filtro
    
    % Filtragem e Oversampling
    y_filtered = rcosflt(y_polar, RB_f, Fa, 'filter', num)'; % rcosflt já faz upsampling/filtragem
else % MATLAB
    % Projeto do filtro Raised-Cosine (MATLAB: usa rcosdesign)
    % 6 é o atraso em símbolos
    h = rcosdesign(r, 6, sps);
    
    % Upsampling e Filtragem FIR (upfirdn)
    y_filtered = upfirdn(y_polar, h, sps);
end

% Ajuste de comprimento (opcional, dependendo do rcosflt/upfirdn)
% Se necessário, o sinal pode ser normalizado ou truncado aqui.


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% 5. Modulacao em Banda Passante (BPSK)            %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
t = (0:length(y_filtered) - 1)/Fa;
y = y_filtered .* cos(2 * pi * Fp * t); % Modulação BPSK

% Retorna o vetor de bits codificado e enquadrado para inspeção
msg_encoded = msg_framed;

end


function coded_bits = hamming_encode(msg_bits)
% hamming_encode - Codifica um vetor de bits usando Hamming (7,4)k = 4;  % Número de bits de dados
n = 7;  % Tamanho do bloco codificado
len_msg = length(msg_bits);
num_blocks = ceil(len_msg / k);

% Padding (preenche com zeros se necessário para completar o último bloco)
padded_bits = [msg_bits zeros(1, num_blocks * k - len_msg)];

coded_bits = [];

for i = 1:num_blocks
    data_block = padded_bits((i-1)*k + 1 : i*k);
    % A função 'encode' no Octave pode retornar um vetor coluna, por isso a transposição (').
    coded_block = encode(data_block, n, k, 'hamming')';
    coded_bits = [coded_bits coded_block];
end

end

File: modem_acustico_python/transmissor.py
import numpy as np

# ====================== Funções auxiliares ======================

def bits_from_int(value, width=8):
    """
    Converte inteiro em lista de bits LSB-first
    (equivalente ao de2bi(..., 'right-msb') do MATLAB).
    """
    return [(value >> i) & 1 for i in range(width)]


# ====================== Hamming (7,4) ======================

def hamming74_encode(msg_bits):
    """
    Hamming (7,4) encoder.
    Entrada: vetor 1D de bits (0/1).
    Saída: numpy array de bits codificados.
    Ordem em cada bloco: [p1 p2 d1 p3 d2 d3 d4]
    """
    msg_bits = np.array(msg_bits, dtype=int).flatten()
    k = 4
    n = 7
    len_msg = len(msg_bits)
    num_blocks = int(np.ceil(len_msg / k))

    # Padding com zeros se necessário
    padded = np.concatenate([msg_bits,
                             np.zeros(num_blocks * k - len_msg, dtype=int)])

    coded_blocks = []
    for i in range(num_blocks):
        d1, d2, d3, d4 = padded[i*k:(i+1)*k]
        # Paridades para [p1 p2 d1 p3 d2 d3 d4]
        p1 = (d1 ^ d2 ^ d4) & 1      # posições 1,3,5,7
        p2 = (d1 ^ d3 ^ d4) & 1      # posições 2,3,6,7
        p3 = (d2 ^ d3 ^ d4) & 1      # posições 4,5,6,7
        block = np.array([p1, p2, d1, p3, d2, d3, d4], dtype=int)
        coded_blocks.append(block)

    return np.concatenate(coded_blocks)


# ====================== Filtro RRC (equivalente ao rcosdesign) ======================

def rrcosfilter(span, sps, rolloff):
    """
    Gera um filtro raiz cosseno levantado (RRC) similar ao rcosdesign do MATLAB.
    span: atraso em símbolos (lado inteiro, total = span símbolos)
    sps: amostras por símbolo
    rolloff: fator de roll-off (0 < rolloff <= 1)
    Retorna: (t, h)
    """
    N = span * sps
    t = np.arange(-N/2, N/2 + 1) / float(sps)  # em tempos de símbolo
    h = np.zeros_like(t, dtype=float)

    for i, ti in enumerate(t):
        if abs(ti) < 1e-8:
            h[i] = 1.0 - rolloff + (4 * rolloff / np.pi)
        elif abs(abs(4 * rolloff * ti) - 1.0) < 1e-8:
            # ti = ± Ts / (4*rolloff)
            h[i] = (rolloff / np.sqrt(2)) * (
                ((1 + 2/np.pi) * np.sin(np.pi / (4 * rolloff))) +
                ((1 - 2/np.pi) * np.cos(np.pi / (4 * rolloff)))
            )
        else:
            num = (np.sin(np.pi * ti * (1 - rolloff)) +
                   4 * rolloff * ti * np.cos(np.pi * ti * (1 + rolloff)))
            den = (np.pi * ti * (1 - (4 * rolloff * ti)**2))
            h[i] = num / den

    # Normaliza energia
    h = h / np.sqrt(np.sum(h**2))
    return t, h


# ====================== TRANSMISSOR ======================

def transmissor(msg, RB, Fp, Fa):
    """
    Transmissor BPSK com:
      - Byte de comprimento
      - Hamming(7,4)
      - PREAMBLE + SFD + payload + PREAMBLE
      - Pulso RRC
    msg : string de texto
    RB  : taxa de bits (bits/s)
    Fp  : frequência da portadora (Hz)
    Fa  : frequência de amostragem (Hz)
    Retorna:
      y          : sinal modulado (float)
      msg_framed : bits após enquadramento
    """

    # SFD (32 bits)
    SFD = np.array([
        1,1,0,0,1,1,1,0,0,0,1,1,1,1,0,0,
        0,0,1,1,1,0,0,0,1,1,0,0,1,0,1,0
    ], dtype=int)

    # PREAMBLE original: [ones(1,10) upsample(ones(1,15),2)]
    pre_sync = np.zeros(15*2, dtype=int)
    pre_sync[::2] = 1          # [1 0 1 0 ...]
    PREAMBLE = np.concatenate([np.ones(10, dtype=int), pre_sync])

    # 1. Texto -> bits
    if isinstance(msg, str) and len(msg) > 0:
        l = len(msg)
        msg_dec = [ord(c) for c in msg]

        # Byte de comprimento
        l_bits = np.array(bits_from_int(l, 8), dtype=int)

        # Bits dos dados (8 bits por caractere)
        msg_data_bits = []
        for val in msg_dec:
            msg_data_bits.extend(bits_from_int(val, 8))
        msg_data_bits = np.array(msg_data_bits, dtype=int)

        # [Byte_Comprimento | Dados]
        msg_bits_raw = np.concatenate([l_bits, msg_data_bits])

        # Hamming(7,4)
        msg_encoded = hamming74_encode(msg_bits_raw)
    elif isinstance(msg, (list, np.ndarray)):
        msg_encoded = np.array(msg, dtype=int).flatten()
    else:
        raise ValueError('Input "engenharia" deve ser uma string ou vetor de bits.')

    print(f"Tamanho do payload (bits) antes do enquadramento: {len(msg_encoded)}")

    # 2. Enquadramento
    msg_framed = np.concatenate([PREAMBLE, SFD, msg_encoded, PREAMBLE])

    # 3. Polar: 0 -> -1, 1 -> +1
    y_polar = 2 * msg_framed - 1

    print(f"Tamanho do quadro: {len(y_polar)} símbolos")

    # 4. Pulso RRC
    r = 0.5
    sps = int(np.floor(Fa / RB))  # samples per symbol
    if sps < 1:
        raise ValueError("Fa/RB precisa ser >= 1 para ao menos 1 amostra por símbolo")

    span = 6
    _, h = rrcosfilter(span, sps, r)

    # Upsampling
    up = np.zeros(len(y_polar) * sps)
    up[::sps] = y_polar

    # Filtragem
    y_filtered = np.convolve(up, h, mode='full')

    # 5. Modulação BPSK
    t = np.arange(len(y_filtered)) / float(Fa)
    y = y_filtered * np.cos(2 * np.pi * Fp * t)

    return y, msg_framed
